### DISCUSSION TITLE
Persistence, Authentication and other "Essentials"

### DISCUSSION THREAD

**Discussion by lylerolleman** [AvaloniaUI/Avalonia#12621] [General]
I'm wondering what level of support exists (formal or 3rd party) for a number of common, generally platform specific features. (that is, Xamarin.Essentials-esque kind of stuff)

Specific items I'm interested in are

- Key-value persistence (with or without encrypted store access)
- Cross platform file system access (can't actually find documentation for this one, which surprises me)
- Web Authenticator (that is, open a browser window to authenticate with OAuth on non-Web platforms)
- Biometric support (on supported platforms)
- Requesting permissions (again, where relevant)

I am fairly confident I could achieve most of these by DI-ing platform specific implementations (or just using a simple database on the filesystem in the case of key-value persistence) but I'd rather not re-invent the wheel if I don't have to.

**Comment by timunie** [comment]
Xamarin.Essentails and Maui.Essentials can be used with Avalonia (most of them at least)... In the long term we may want to extend also our own set of Essentails, due to lack of some platforms and features (https://github.com/AvaloniaUI/Avalonia.Essentials).

**Reply by maxkatz6** [reply]
Update: Maui.Essentials is now detached from the Maui workloads, starting with .NET 8. There is no need for another avalonia specific package. 
Keep in mind though, Essentials package is platform limited, and mostly useful only for mobile.

**Comment by timunie** [comment]
> Cross platform file system access (can't actually find documentation for this one, which surprises me)

Use Toplevel.StorageProvider (see: https://docs.avaloniaui.net/docs/next/concepts/services/storage-provider/)

**Reply by lylerolleman** [reply]
Outstanding, looks like just what I've been looking for

Is there a nuget for Avalonia.Essentials or is it just build from source for now?

**Reply by timunie** [reply]
source only iirc

**Comment by schauerte** [comment]
I am confused, what the current state on this topic is.

I've created a new Avalonia 11.0.6 cross-platform application (the avalonia.xplat template with net8.0) and tried to query the battery information for testing.

### Xamarin.Essentials
After adding Xamarin.Essentials (to all projects), the following runtime exception is thrown when running in the iOS project (which references the NuGet package directly):
> This functionality is not implemented in the portable version of this assembly. You should reference the NuGet package from your main application project in order to reference the platform-specific implementation. (Xamarin.Essentials.NotImplementedInReferenceAssemblyException)

### Microsoft.Maui.Essentials
I was sceptical, because it references Microsoft.Maui.Graphics but since James Montemagno [says](https://devblogs.microsoft.com/xamarin/xamarin-essentials-1-7-and-introducing-net-maui-essentials)
"_.NET MAUI is the future of Xamarin.Forms and .NET MAUI Essentials is the future of Xamarin.Essentials._" I gave it a try. 

Result: 
> System.InvalidOperationException: Infinite layout loop detected

### Avalonia.Essentails
Seems deprecated: "_This repository has been archived by the owner on Dec 23, 2023. It is now read-only._"

When Xamarin./Maui.Essentials could be used, why does Avalonia.Essentials exist? What should be used? Did I miss anything?

**Reply by maxkatz6** [reply]
Use MAUI.Essentials for mobile. You can use it directly from nuget (.net 8+). Or via Avalonia.Maui integration package. https://github.com/AvaloniaUI/AvaloniaMauiHybrid

**Reply by maxkatz6** [reply]
Keep in mind, maui essentials package is mostly about Mobile support. As MAUI in general is.

**Reply by schauerte** [reply]
Thanks a lot for your replies!

I've followed the instructions (from the linked repo) and ended up with the same exception (while simulating an IPhone 15 with 17.x). So I cloned your repo and run the sample application with success (using a simulated iPad by accident).

Turned out: your sample runs with all simulators I've tried while my project does only work with some (e.g. iPad/IPhone SE) but not all (e.g. IPhone 15).

I don't see a significant difference between projects (and have no idea how to hunt this down - because I really don't know how the moving parts fit together). Maybe I will try again with a new sample application but for now I've got the impression that Xamarin/MAUI may still not be reliable options right now.

**Reply by maxkatz6** [reply]
I can't tell what's wrong in your situation. But if you want to avoid MAUI dependency, you can call Battery APIs directly.
We have a little bit documentation on conditional compilation - https://docs.avaloniaui.net/docs/guides/platforms/platform-specific-code/dotnet#conditional-compilation(as well as platform specific projects on the same page below). Which allows you to call platform APIs from your projects. Pretty much any projects, not only Avalonia.

**Reply by maxkatz6** [reply]
On Android you would need to use [BatteryManager APIs](https://stackoverflow.com/a/38509011). Something like:
```
var bm = Application.Context.GetSystemService(Context.BatteryService).JavaCast<BatteryManager>();
int batLevel = bm.GetIntProperty((int)BatteryProperty.Capacity);
```

And on iOS - [UIDevice](https://stackoverflow.com/a/27475577):
```
UIDevice.CurrentDevice.BatteryMonitoringEnabled = true; // probably should be called once in app startup
var val = UIDevice.CurrentDevice.BatteryLevel;
```

Note, I haven't tested these, just googled quickly, and adapted to C# code.

**Reply by maxkatz6** [reply]
Updated Avalonia.Essentials readme to make it less confusing.

**Reply by schauerte** [reply]
The updated readme is a big improvement. Thank you for that - and also for the code snippets (I was going to try copy and paste some of the essentials source-code as a last resort).

After setting up a new project and adding only `<UseMauiEssentials>true</UseMauiEssentials>` I was able to reference and use Microsoft.Maui.Essentials without issue (running on simulated devices that were broken earlier) - at least to query the battery state, but this was just the first quick test.
